language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/InterfaceCompanion.java

```diff
@@ -23,8 +30,40 @@
    * and <em>should not be referenced directly</em>. Applications should refer to the {@code
    * INTERFACE} field on generated code for Daml interfaces instead.
    */
-  protected InterfaceCompanion(Identifier templateId, ValueDecoder<View> valueDecoder) {
+  protected InterfaceCompanion(
+      Identifier templateId, Function<String, Id> newContractId, ValueDecoder<View> valueDecoder) {
     super(templateId);
+    this.newContractId = newContractId;
     this.valueDecoder = valueDecoder;
   }
+
+  private Contract<Id, View> fromIdAndRecord(
+      String contractId,
+      Map<Identifier, DamlRecord> interfaceViews,
+      Optional<String> agreementText,
+      Set<String> signatories,
+      Set<String> observers)
+      throws IllegalArgumentException {
+    Optional<DamlRecord> maybeRecord = Optional.ofNullable(interfaceViews.get(TEMPLATE_ID));
+    Optional<DamlRecord> maybeFailedRecord = Optional.ofNullable(interfaceViews.get(TEMPLATE_ID));
+    Id id = newContractId.apply(contractId);
+    if (maybeFailedRecord.isPresent()) {
+      throw new IllegalArgumentException("Failed interface view for " + TEMPLATE_ID);
+    } else if (maybeRecord.isPresent()) {
+      View view = valueDecoder.decode(maybeRecord.get());
+      return new ContractWithInterfaceView<>(id, view, agreementText, signatories, observers);
+    } else {
+      throw new IllegalArgumentException("interface view of " + TEMPLATE_ID + " not found.");
```

Comment:
Use `map` and `orElseThrow` here, not `isPresent` and `get`.

Reply:
By the way, if you have an `interfaceViews` entry you will not have a `failedInterfaceViews` entry. So you might as well check the former first, and only bother checking the latter if you want to have a different exception message.